tools/create_snapshot_artifact/main.py

```diff
@@ -0,0 +1,305 @@
+            |
+            ...
+    """
+    # Create directory dedicated to store snapshot artifacts for
+    # each guest kernel version.
+    shutil.rmtree(SNAPSHOT_ARTIFACTS_ROOT_DIR, ignore_errors=True)
+    os.mkdir(SNAPSHOT_ARTIFACTS_ROOT_DIR)
+
+    root_path = tempfile.mkdtemp(
+        prefix=MicrovmBuilder.ROOT_PREFIX,
+        dir=f"{DEFAULT_TEST_SESSION_ROOT_PATH}")
+
+    # Compile new-pid cloner helper.
+    bin_cloner_path = compile_file(
+        file_name='newpid_cloner.c',
+        bin_name='newpid_cloner',
+        dest_path=root_path
+    )
+
+    # Fetch kernel and rootfs artifacts from S3 bucket.
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    kernel_artifacts = ArtifactSet(artifacts.kernels())
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword="ubuntu"))
+
+    # Create a test context.
+    test_context = TestContext()
+    test_context.custom = {
+        'bin_cloner_path': bin_cloner_path,
+        'session_root_path': root_path
+    }
+
+    # Create the test matrix.
+    test_matrix = TestMatrix(context=test_context,
+                             artifact_sets=[
+                                 kernel_artifacts,
+                                 disk_artifacts
+                             ])
+
+    test_matrix.run_test(create_snapshots)
+
+
+def create_snapshots(context):
+    """Snapshot microVM built from vm configuration file."""
+    vm = setup_vm(context)
+
+    # Get ssh key from read-only artifact.
+    ssh_key = context.disk.ssh_key()
+    ssh_key.download(vm.path)
+    vm.ssh_config['ssh_key_path'] = ssh_key.local_path()
+    os.chmod(vm.ssh_config['ssh_key_path'], 0o400)
+
+    _configure_vm_from_json(vm, VM_CONFIG_FILE)
+    configure_network_interfaces(vm)
+    vm.spawn()
+
+    # Ensure the microVM has started.
+    response = vm.machine_cfg.get()
+    assert vm.api_session.is_status_ok(response.status_code)
+    assert vm.state == "Running"
+
+    # Populate MMDS.
+    data_store = {
+        'latest': {
+            'meta-data': {
+                'ami-id': 'ami-12345678',
+                'reservation-id': 'r-fea54097',
+                'local-hostname': 'ip-10-251-50-12.ec2.internal',
+                'public-hostname': 'ec2-203-0-113-25.compute-1.amazonaws.com'
+            }
+        }
+    }
+    populate_mmds(vm, data_store)
+
+    # Iterate and validate connectivity on all ifaces after boot.
+    for iface in net_ifaces:
+        vm.ssh_config['hostname'] = iface.guest_ip
+        ssh_connection = net_tools.SSHConnection(vm.ssh_config)
+        exit_code, _, _ = ssh_connection.execute_command("sync")
+        assert exit_code == 0
+
+    # Validate MMDS.
+    validate_mmds(ssh_connection, data_store)
+
+    # Create a snapshot builder from a microVM.
+    snapshot_builder = SnapshotBuilder(vm)
+
+    # Snapshot the microVM.
+    snapshot = snapshot_builder.create(
+        [vm.rootfs_file],
+        ssh_key,
+        SnapshotType.DIFF,
```

Comment:
Wouldn't we better use FULL snapshots given that we still have not yet merged a solution to [this](https://github.com/firecracker-microvm/firecracker/issues/2802)?

Reply:
Indeed, we are not doing any merging, we should be good.